existential-vs-universal-quantification/article.md

```diff
@@ -0,0 +1,292 @@
+
+```Haskell
+func :: forall a b c m. a -> b -> m c -- OK, it compiles
+```
+Here we introduce four variables.
+All type variables that you introduce in your definition should belong to one `forall` exactly, though they may not belong to the same one. (I will explain it further.)
+
+```haskell
+func :: forall a b c.   a -> b -> m c -- Error, type variable `m` is not introduced
+```
+
+Let's add some constraints.
+For example, we might want to specify the `m` variable here as an instance of `Monad` typeclass.
+
+```Haskell
+func :: forall a b c m. Monad m => a -> b -> m c
+```
+You might want to ask, what's so special about this syntax?
+Nothing changed since we added the quantifier because it was already there!
+However, it allows us to use new tricks.
+
+### Examples
+
+#### Order of type variables
+
+First of all, you can now change the order of type variables.
+Let's imagine you want to define a function with 10 type variables (yeah, there are real-life [examples](https://hackage.haskell.org/package/leancheck-0.9.10/docs/Test-LeanCheck-Utils-TypeBinding.html#v:-45--62--62--62--62--62--62--62--62--62--62--62--62-:) of such functions), and you know that when you use it, you want to specify type of the last argument explicitly.
+Here, and in the future examples we will use the `TypeApplication` extension, which allows you to pass types as function arguments, using `@` sign (If specifying concrete type is unnecessary you can use type wildcard `@_`).
+By default, the order of type variables, which you assign types to, is the same as the order in which they are introduced in the function definition.
+Without `forall`, you would write something like this:
+
+```Haskell
+veryLongFunction :: a -> b -> c -> d -> e -> f -> g -> h -> i -> j
+veryLongFunction = ...
+
+func = veryLongFunction @_ @_ @_ @_ @_ @_ @_ @_ @_ @Integer ...
+```
+Quite long, right?
+However, by using explicit `forall`, this can be simplified:
+
+```Haskell
+veryLongFunction :: forall j a b c d e f g h i. a -> b -> c -> d -> e -> f -> g -> h -> i -> j
+veryLongFunction = ...
+
+func = veryLongFunction @Integer ...
+```
+
+Since `j` is the first type variable in the declaration of `veryLongFunction` you can specify only it.
+This is useful not only in such big examples but also when your function requires any of its arguments to be specified on usage.
+
+### Support of other extensions
+
+`ExplicitForAll` shines most when you need to enable other extensions – some of them just won't work without the `forall` quantifier. :)
+The most common one is `ScopedTypeVariables`, which allows you to pass a type variable from an outer function to the one defined in a where clause.
+
+```haskell
+example :: a -> [a] -> [a]
+example x rest = pair ++ rest
+  where
+    pair :: [a]
+    pair = [x, x]
+```
+
+This will cause an error without `ScopedTypeVariables`, because compiler thinks that `a` in `example`'s signature and `a` in `double`'s are two different types.
```

Comment:
>because compiler thinks that `a` in `example`'s signature and `a` in `double`'s are two different types.

AFAIU, the error happens because the compiler takes `pair` function to be a polymorphic one, which should be able to return a list of values of *any* type. But its implementation `[x, x]` can only be a single concrete type, as dictated by the concrete type variable from the `example` function. 

So I think this can be easily corrected by saying "two different type variables" instead of "two different types".

Reply:
@sras I expanded on the explanation of `ScopedTypeVariables` here: https://github.com/serokell/blog-posts/pull/4/commits/aabbf1fb65912d7ee8dcfc2f2698b6f8a4f52719